internal/mailer/validate.go

```diff
@@ -0,0 +1,290 @@
+package mailer
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"net/mail"
+	"strings"
+	"time"
+
+	"github.com/supabase/auth/internal/conf"
+	"golang.org/x/sync/errgroup"
+)
+
+var invalidEmailMap = map[string]bool{
+
+	// People type these often enough to be special cased.
+	"test@gmail.com":    true,
+	"example@gmail.com": true,
+	"someone@gmail.com": true,
+	"test@email.com":    true,
+}
+
+var invalidHostSuffixes = []string{
+
+	// These are a directly from Section 2 of RFC2606[1].
+	//
+	// [1] https://www.rfc-editor.org/rfc/rfc2606.html#section-2
+	".test",
+	".example",
+	".invalid",
+	".local",
+	".localhost",
+}
+
+var invalidHostMap = map[string]bool{
+
+	// These exist here too for when they are typed as "test@test"
+	"test":      true,
+	"example":   true,
+	"invalid":   true,
+	"local":     true,
+	"localhost": true,
+
+	// These are commonly typed and have DNS records which cause a
+	// large enough volume of bounce backs to special case.
+	"test.com":    true,
+	"example.com": true,
+	"example.net": true,
+	"example.org": true,
+
+	// Hundreds of typos per day for this.
+	"gamil.com": true,
+
+	// These are not email providers, but people often use them.
+	"anonymous.com": true,
+	"email.com":     true,
+}
+
+const (
+	validateEmailTimeout = 2 * time.Second
+)
+
+var (
+	// We use the default resolver for this.
+	validateEmailResolver net.Resolver
+)
+
+var (
+	ErrInvalidEmailAddress = errors.New("invalid_email_address")
+	ErrInvalidEmailFormat  = errors.New("invalid_email_format")
+	ErrInvalidEmailDNS     = errors.New("invalid_email_dns")
+)
+
+type EmailValidator struct {
+	extended   bool
+	serviceURL string
+	serviceKey string
+}
+
+func newEmailValidator(mc conf.MailerConfiguration) *EmailValidator {
+	return &EmailValidator{
+		extended:   mc.EmailValidationExtended,
+		serviceURL: mc.EmailValidationServiceURL,
+		serviceKey: mc.EmailValidationServiceKey,
+	}
+}
+
+func (ev *EmailValidator) isExtendedDisabled() bool { return !ev.extended }
```

Comment:
i have a slight preference for renaming this to `isExtendedEnabled` 
```suggestion
func (ev *EmailValidator) isExtendedEnabled() bool { return ev.extended }
```

Reply:
just so that it's easier to not have to think about double negatives